src/Orleans.Core.Abstractions/Core/Grain.cs

```diff
@@ -196,39 +202,48 @@ protected Task<List<IGrainReminder>> GetReminders()
         protected void DeactivateOnIdle()
         {
             EnsureRuntime();
-            Runtime.DeactivateOnIdle(this);
+            Runtime.DeactivateOnIdle(_grainContext ?? RuntimeContext.Current);
         }
 
         /// <summary>
         /// Delay Deactivation of this activation at least for the specified time duration.
         /// A positive <c>timeSpan</c> value means “prevent GC of this activation for that time span”.
         /// A negative <c>timeSpan</c> value means “cancel the previous setting of the DelayDeactivation call and make this activation behave based on the regular Activation Garbage Collection settings”.
-        /// DeactivateOnIdle method would undo / override any current “keep alive” setting, 
+        /// DeactivateOnIdle method would undo / override any current “keep alive” setting,
         /// making this grain immediately available for deactivation.
         /// </summary>
         protected void DelayDeactivation(TimeSpan timeSpan)
         {
             EnsureRuntime();
-            Runtime.DelayDeactivation(this, timeSpan);
+            Runtime.DelayDeactivation(_grainContext ?? RuntimeContext.Current, timeSpan);
         }
 
         /// <summary>
         /// This method is called at the end of the process of activating a grain.
         /// It is called before any messages have been dispatched to the grain.
         /// For grains with declared persistent state, this method is called after the State property has been populated.
         /// </summary>
-        public virtual Task OnActivateAsync()
-        {
-            return Task.CompletedTask;
-        }
+        /// <param name="cancellationToken">A cancellation token which signals when activation is being canceled.</param>
+        public virtual Task OnActivateAsync(CancellationToken cancellationToken) => OnActivateAsync();
+
+        /// <summary>
+        /// This method is called at the end of the process of activating a grain.
+        /// It is called before any messages have been dispatched to the grain.
+        /// For grains with declared persistent state, this method is called after the State property has been populated.
+        /// </summary>
+        public virtual Task OnActivateAsync() => Task.CompletedTask;
```

Comment:
I am afraid that having two different overridable `OnActivateAsync` method will be confusing for people. (and it's the same for `OnDeactivateAsync`)

Reply:
I can delete these no-arg overloads